### ISSUE TITLE
grad not found after "breaking change"

### ISSUE THREAD

**Issue by aespielberg** [yuanming-hu/taichi#321]
taichi 0.3.7, Ubuntu 16.04, Anaconda 3.7.4

I'm running my code after updating, and I made the changes suggested in the log.  This is the output I get:

```
$ python3 test_mpm_simulator.py 
[Release mode]
[T 12/23/19 21:06:28.326] [logging.cpp:Logger@68] Taichi core started. Thread ID = 41341
[Taichi version 0.3.7, cuda 10.0, commit dbe66532]
add actuator with id  0
add actuator with id  1
add actuator with id  2
add actuator with id  3
Using CUDA Device [0]: Quadro RTX 8000
Device Compute Capability: 7.5
[I 12/23/19 21:06:30.505] [taichi_llvm_context.cpp:TaichiLLVMContext@59] Creating llvm context for arch: x86_64
[I 12/23/19 21:06:30.638] [/home/aespielberg2/anaconda3/lib/python3.7/site-packages/taichi/lang/impl.py:materialize@111] Materializing layout...
[W 12/23/19 21:06:30.639] [snode.cpp:create_node@44] Non-power-of-two node size 4000 promoted to 4096.
[W 12/23/19 21:06:30.639] [snode.cpp:create_node@44] Non-power-of-two node size 4000 promoted to 4096.
[W 12/23/19 21:06:30.639] [snode.cpp:create_node@44] Non-power-of-two node size 4000 promoted to 4096.
[W 12/23/19 21:06:30.639] [snode.cpp:create_node@44] Non-power-of-two node size 3100 promoted to 4096.
[W 12/23/19 21:06:30.639] [snode.cpp:create_node@44] Non-power-of-two node size 4000 promoted to 4096.
[W 12/23/19 21:06:30.639] [snode.cpp:create_node@44] Non-power-of-two node size 3100 promoted to 4096.
[W 12/23/19 21:06:30.640] [snode.cpp:create_node@44] Non-power-of-two node size 4000 promoted to 4096.
[W 12/23/19 21:06:30.640] [snode.cpp:create_node@44] Non-power-of-two node size 4000 promoted to 4096.
[W 12/23/19 21:06:30.641] [snode.cpp:create_node@44] Non-power-of-two node size 4000 promoted to 4096.
[W 12/23/19 21:06:30.641] [snode.cpp:create_node@44] Non-power-of-two node size 4000 promoted to 4096.
[W 12/23/19 21:06:30.641] [snode.cpp:create_node@44] Non-power-of-two node size 3100 promoted to 4096.
[W 12/23/19 21:06:30.641] [snode.cpp:create_node@44] Non-power-of-two node size 100 promoted to 128.
[I 12/23/19 21:06:30.927] [codegen_llvm_x86.cpp:global_optimize_module_x86_64@93] Global optimization time: 257.589 ms
[I 12/23/19 21:06:30.932] [struct_llvm.cpp:operator()@268] Allocating data structure of size 2956104296
Initializing runtime with 108 elements
Runtime initialized.
[I 12/23/19 21:06:30.999] [taichi_llvm_context.cpp:TaichiLLVMContext@59] Creating llvm context for arch: cuda
[I 12/23/19 21:06:31.383] [codegen_llvm_x86.cpp:global_optimize_module_x86_64@93] Global optimization time: 52.0749 ms
[I 12/23/19 21:06:31.588] [codegen_llvm_x86.cpp:global_optimize_module_x86_64@93] Global optimization time: 74.398 ms
[I 12/23/19 21:06:31.786] [codegen_llvm_x86.cpp:global_optimize_module_x86_64@93] Global optimization time: 74.842 ms
[I 12/23/19 21:06:31.979] [codegen_llvm_x86.cpp:global_optimize_module_x86_64@93] Global optimization time: 59.2408 ms
[I 12/23/19 21:06:32.142] [codegen_llvm_x86.cpp:global_optimize_module_x86_64@93] Global optimization time: 62.947 ms
[I 12/23/19 21:06:32.337] [codegen_llvm_x86.cpp:global_optimize_module_x86_64@93] Global optimization time: 64.8839 ms
[I 12/23/19 21:06:32.524] [codegen_llvm_x86.cpp:global_optimize_module_x86_64@93] Global optimization time: 60.0619 ms
[I 12/23/19 21:06:32.678] [codegen_llvm_x86.cpp:global_optimize_module_x86_64@93] Global optimization time: 57.714 ms
[I 12/23/19 21:06:33.412] [codegen_llvm_x86.cpp:global_optimize_module_x86_64@93] Global optimization time: 70.39 ms
[I 12/23/19 21:06:33.859] [codegen_llvm_x86.cpp:global_optimize_module_x86_64@93] Global optimization time: 41.8689 ms
[I 12/23/19 21:06:34.036] [codegen_llvm_x86.cpp:global_optimize_module_x86_64@93] Global optimization time: 60.6711 ms
[I 12/23/19 21:06:34.227] [codegen_llvm_x86.cpp:global_optimize_module_x86_64@93] Global optimization time: 68.4211 ms
[I 12/23/19 21:06:34.445] [codegen_llvm_x86.cpp:global_optimize_module_x86_64@93] Global optimization time: 74.316 ms
num particles  3100
[I 12/23/19 21:06:34.808] [codegen_llvm_x86.cpp:global_optimize_module_x86_64@93] Global optimization time: 65.99 ms
[I 12/23/19 21:06:35.055] [codegen_llvm_x86.cpp:global_optimize_module_x86_64@93] Global optimization time: 55.2559 ms
[I 12/23/19 21:06:35.283] [codegen_llvm_x86.cpp:global_optimize_module_x86_64@93] Global optimization time: 67.723 ms
[I 12/23/19 21:06:35.453] [codegen_llvm_x86.cpp:global_optimize_module_x86_64@93] Global optimization time: 46.1121 ms
[I 12/23/19 21:06:35.560] [codegen_llvm_x86.cpp:global_optimize_module_x86_64@93] Global optimization time: 39.5169 ms
[I 12/23/19 21:06:35.680] [codegen_llvm_x86.cpp:global_optimize_module_x86_64@93] Global optimization time: 46.5329 ms
[I 12/23/19 21:06:35.846] [llvm_jit_ptx.cpp:compile@190] PTX size: 3.76KB
[I 12/23/19 21:06:36.031] [llvm_jit_ptx.cpp:compile@190] PTX size: 3.32KB
[I 12/23/19 21:06:36.212] [llvm_jit_ptx.cpp:compile@190] PTX size: 3.40KB
[I 12/23/19 21:06:36.406] [llvm_jit_ptx.cpp:compile@190] PTX size: 2.87KB
[I 12/23/19 21:06:36.573] [llvm_jit_ptx.cpp:compile@190] PTX size: 2.87KB
[I 12/23/19 21:06:36.768] [llvm_jit_ptx.cpp:compile@190] PTX size: 3.78KB
[I 12/23/19 21:06:36.972] [llvm_jit_ptx.cpp:compile@190] PTX size: 3.78KB
[I 12/23/19 21:06:37.181] [llvm_jit_ptx.cpp:compile@190] PTX size: 3.72KB
[I 12/23/19 21:06:37.441] [llvm_jit_ptx.cpp:compile@190] PTX size: 24.34KB
[I 12/23/19 21:06:37.654] [llvm_jit_ptx.cpp:compile@190] PTX size: 3.55KB
[I 12/23/19 21:06:37.817] [llvm_jit_ptx.cpp:compile@190] PTX size: 3.66KB
[I 12/23/19 21:06:37.994] [llvm_jit_ptx.cpp:compile@190] PTX size: 3.69KB
[I 12/23/19 21:06:38.170] [llvm_jit_ptx.cpp:compile@190] PTX size: 3.82KB
[I 12/23/19 21:06:38.374] [llvm_jit_ptx.cpp:compile@190] PTX size: 4.17KB
[I 12/23/19 21:06:38.562] [llvm_jit_ptx.cpp:compile@190] PTX size: 2.92KB
[I 12/23/19 21:06:38.719] [llvm_jit_ptx.cpp:compile@190] PTX size: 0.75KB
[I 12/23/19 21:06:38.869] [codegen_llvm_x86.cpp:global_optimize_module_x86_64@93] Global optimization time: 41.7738 ms
[I 12/23/19 21:06:38.998] [codegen_llvm_x86.cpp:global_optimize_module_x86_64@93] Global optimization time: 30.8552 ms
{}
[I 12/23/19 21:06:39.011] [/home/aespielberg2/anaconda3/lib/python3.7/site-packages/taichi/lang/kernel.py:materialize@175] Compiling kernel clear_grid_0_...
[I 12/23/19 21:06:39.225] [llvm_jit_ptx.cpp:compile@190] PTX size: 3.63KB
[I 12/23/19 21:06:39.375] [codegen_llvm_x86.cpp:global_optimize_module_x86_64@93] Global optimization time: 64.187 ms
[I 12/23/19 21:06:39.503] [codegen_llvm_x86.cpp:global_optimize_module_x86_64@93] Global optimization time: 50.715 ms
{}
[I 12/23/19 21:06:39.514] [/home/aespielberg2/anaconda3/lib/python3.7/site-packages/taichi/lang/kernel.py:materialize@175] Compiling kernel compute_v_avg_0_...
[I 12/23/19 21:06:39.703] [llvm_jit_ptx.cpp:compile@190] PTX size: 1.44KB
{}
[I 12/23/19 21:06:39.707] [/home/aespielberg2/anaconda3/lib/python3.7/site-packages/taichi/lang/kernel.py:materialize@175] Compiling kernel nn1_0_...
[I 12/23/19 21:06:39.900] [llvm_jit_ptx.cpp:compile@190] PTX size: 2.84KB
{}
[I 12/23/19 21:06:39.904] [/home/aespielberg2/anaconda3/lib/python3.7/site-packages/taichi/lang/kernel.py:materialize@175] Compiling kernel nn2_0_...
[I 12/23/19 21:06:40.669] [llvm_jit_ptx.cpp:compile@190] PTX size: 9.08KB
{}
[I 12/23/19 21:06:40.673] [/home/aespielberg2/anaconda3/lib/python3.7/site-packages/taichi/lang/kernel.py:materialize@175] Compiling kernel p2g_0_...
[I 12/23/19 21:06:41.396] [llvm_jit_ptx.cpp:compile@190] PTX size: 12.18KB
{}
[I 12/23/19 21:06:41.399] [/home/aespielberg2/anaconda3/lib/python3.7/site-packages/taichi/lang/kernel.py:materialize@175] Compiling kernel grid_op_0_...
[I 12/23/19 21:06:41.781] [llvm_jit_ptx.cpp:compile@190] PTX size: 4.82KB
{}
[I 12/23/19 21:06:41.783] [/home/aespielberg2/anaconda3/lib/python3.7/site-packages/taichi/lang/kernel.py:materialize@175] Compiling kernel g2p_0_...
[I 12/23/19 21:06:42.256] [llvm_jit_ptx.cpp:compile@190] PTX size: 10.26KB
{}
{}
(these braces continue on for hundreds of lines, so I'm skipping them here)

[I 12/23/19 21:06:45.982] [codegen_llvm_x86.cpp:global_optimize_module_x86_64@93] Global optimization time: 29.2981 ms
[I 12/23/19 21:06:46.106] [codegen_llvm_x86.cpp:global_optimize_module_x86_64@93] Global optimization time: 39.5629 ms
{}
[I 12/23/19 21:06:46.116] [/home/aespielberg2/anaconda3/lib/python3.7/site-packages/taichi/lang/kernel.py:materialize@175] Compiling kernel compute_x_avg_0_...
[I 12/23/19 21:06:46.289] [llvm_jit_ptx.cpp:compile@190] PTX size: 1.26KB
{}
[I 12/23/19 21:06:46.291] [/home/aespielberg2/anaconda3/lib/python3.7/site-packages/taichi/lang/kernel.py:materialize@175] Compiling kernel compute_loss_0_...
[I 12/23/19 21:06:46.437] [llvm_jit_ptx.cpp:compile@190] PTX size: 0.68KB
{}
[I 12/23/19 21:06:46.439] [/home/aespielberg2/anaconda3/lib/python3.7/site-packages/taichi/lang/kernel.py:materialize@175] Compiling kernel compute_total_loss_0_...
[I 12/23/19 21:06:46.575] [llvm_jit_ptx.cpp:compile@190] PTX size: 0.80KB
{}
[I 12/23/19 21:06:46.586] [/home/aespielberg2/anaconda3/lib/python3.7/site-packages/taichi/lang/kernel.py:materialize@175] Compiling kernel compute_total_loss_0__grad...
[I 12/23/19 21:06:46.743] [llvm_jit_ptx.cpp:compile@190] PTX size: 0.79KB
{}
[I 12/23/19 21:06:46.746] [/home/aespielberg2/anaconda3/lib/python3.7/site-packages/taichi/lang/kernel.py:materialize@175] Compiling kernel compute_loss_0__grad...
[I 12/23/19 21:06:46.883] [llvm_jit_ptx.cpp:compile@190] PTX size: 0.78KB
{}
[I 12/23/19 21:06:46.886] [/home/aespielberg2/anaconda3/lib/python3.7/site-packages/taichi/lang/kernel.py:materialize@175] Compiling kernel compute_x_avg_0__grad...
[I 12/23/19 21:06:47.065] [llvm_jit_ptx.cpp:compile@190] PTX size: 1.32KB
{}
{}
{}
Traceback (most recent call last):
  File "test_mpm_simulator.py", line 156, in <module>
    main()
  File "test_mpm_simulator.py", line 130, in main
    sim.forward(sim.steps)
  File "/home/aespielberg2/anaconda3/lib/python3.7/site-packages/taichi/lang/tape.py", line 19, in __exit__
    self.grad()
  File "/home/aespielberg2/anaconda3/lib/python3.7/site-packages/taichi/lang/tape.py", line 29, in grad
    func.grad(*args)
  File "/home/aespielberg2/anaconda3/lib/python3.7/site-packages/taichi/lang/__init__.py", line 181, in decorated
    func(*args, **kwargs)
  File "/home/aespielberg2/ResearchCode/project/mpm_simulator.py", line 657, in advance_grad
    self.g2p.grad(s)
AttributeError: 'function' object has no attribute 'grad'
```

However, this g2p function is written as such:

```
@ti.classkernel
  def g2p(self, f: ti.i32):
    <stuff>
```

The invocation comes from a complex kernel.  The class is decorated as:

```
@ti.data_oriented
class MPMSimulator:
```

Is this user error or is something wrong with taichi?

**Comment by yuanming-hu** [comment]
This is fixed in release v0.3.8.

**Comment by aespielberg** [comment]
Thanks!
